pkg/kube/krt/nestedjoin.go

```diff
@@ -0,0 +1,282 @@
+			return r
+		}
+	}
+	return nil
+}
+
+func (j *nestedjoin[T]) List() []T {
+	var res []T
+	if j.uncheckedOverlap {
+		first := true
+		for _, c := range j.collections.List() {
+			objs := c.List()
+			// As an optimization, take the first (non-empty) result as-is without copying
+			if len(objs) > 0 && first {
+				res = objs
+				first = false
+			} else {
+				// After the first, safely merge into the result
+				res = append(res, objs...)
+			}
+		}
+		return res
+	}
+	var found sets.String
+	first := true
+
+	for _, c := range j.collections.List() {
+		objs := c.List()
+		// As an optimization, take the first (non-empty) result as-is without copying
+		// TODO: Implement custom merge out of the hot path
+		if len(objs) > 0 && first {
+			res = objs
+			first = false
+			found = sets.NewWithLength[string](len(objs))
+			for _, i := range objs {
+				found.Insert(GetKey(i))
+			}
+		} else {
+			// After the first, safely merge into the result
+			for _, i := range objs {
+				key := GetKey(i)
+				if !found.InsertContains(key) {
+					// Only keep it if it is the first time we saw it, as our merging mechanism is to keep the first one
+					res = append(res, i)
+				}
+			}
+		}
+	}
+	return res
+}
+
+// nolint: unused // (not true, its to implement an interface)
+func (j *nestedjoin[T]) index(extract func(o T) []string) kclient.RawIndexer {
+	j.Lock() // Take a write lock since we're also updating the collection change handlers
+	defer j.Unlock()
+	ji := &dynamicJoinIndexer{indexers: make(map[collectionUID]kclient.RawIndexer)}
+	for _, c := range j.collections.List() {
+		ic := c.(internalCollection[T])
+		ji.indexers[ic.uid()] = ic.index(extract)
+	}
+	j.registerCollectionChangeHandlerLocked(func(e collectionChangeEvent[T]) {
+		ji.Lock()
+		defer ji.Unlock()
+		switch e.eventType {
+		case collectionMembershipEventAdd:
+			ji.indexers[e.collectionValue.uid()] = e.collectionValue.index(extract)
+		case collectionMembershipEventDelete:
+			delete(ji.indexers, e.collectionValue.uid())
+		}
+	})
+	return ji
+}
+
+func (j *nestedjoin[T]) Synced() Syncer {
+	return channelSyncer{
+		name:   j.collectionName,
+		synced: j.synced,
+	}
+}
+
+func (j *nestedjoin[T]) HasSynced() bool {
+	return j.syncer.HasSynced()
+}
+
+func (j *nestedjoin[T]) WaitUntilSynced(stop <-chan struct{}) bool {
+	return j.syncer.WaitUntilSynced(stop)
+}
+
+func (j *nestedjoin[T]) Register(f func(o Event[T])) HandlerRegistration {
+	return registerHandlerAsBatched(j, f)
+}
+
+func (j *nestedjoin[T]) RegisterBatch(f func(o []Event[T]), runExistingState bool) HandlerRegistration {
+	j.Lock() // Take a write lock since we're also updating the collection change handlers
+	defer j.Unlock()
+	syncers := make(map[collectionUID]Syncer)
+	removes := map[collectionUID]func(){}
+	for _, c := range j.collections.List() {
+		reg := c.RegisterBatch(f, runExistingState)
+		ic := c.(internalCollection[T])
+		removes[ic.uid()] = reg.UnregisterHandler
+		syncers[ic.uid()] = reg
```

Comment:
nit: why not just store the `HandlerRegistration` then we don't need 2 separate lists

Reply:
🤷🏾 I pretty much just copied the join implementation; not sure why there were separate types. Tests potentially?

Reply:
Oh actually, the lists are necessary because we don't know what the concrete type of the Handler registration is, so we can't dynamically add syncers and removers when collections change